onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -138,7 +138,9 @@ class RemoveDuplicateCastTransformer : public GraphTransformer {
           num_child++;
         }
 
-        if (child_removed == num_child && child_removed > 0) {
+        if (child_removed == num_child && 
+            child_removed > 0 && 
+            graph_outputs.find(node.OutputDefs()[0]) == graph_outputs.end()) {
```

Comment:
Cannot remove the node directly even if all children have been removed - need to make sure the node doesn't contribute to the graph output first before removing it